graph.c

```diff
@@ -0,0 +1,145 @@
+//idea: https://runestone.academy/ns/books/published/pythonds3/Graphs/AnAdjacencyList.html
+//Implementation of simple directed cycled weighted graph
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+// #include "graph_hash_table.c"
+
+#define GRAPH_VERTICES_TABLE_SIZE 10
+
+typedef struct {
+    struct Graph_vertice_neighbor *next;
+    char *name;
+    int weight;
+} Graph_vertice_neighbor;
+
+typedef struct {
+    struct Graph_vertice_neighbor *next;
+    char* name;
+} Graph_vertice;
+
+
+typedef struct {
+    Graph_vertice *vertices_table[GRAPH_VERTICES_TABLE_SIZE];
+} Graph;
+
+Graph graph_create();
+void graph_free(Graph *);
+void graph_list(Graph *);
+
+void graph_add_vertice(Graph*, char *);
+void graph_erase_vertice(Graph*, char *);
+Graph_vertice* graph_get_vertice(char *);
+Graph_vertice** graph_get_vertices(Graph *);
+
+void graph_add_edge(Graph*, char *, char *, int);
+void graph_erase_edge(Graph*, char *, char *);
+
+unsigned hash(char *); //FIXME define in other file
+
+unsigned hash(char *s)
+{
+    unsigned hashval;
+
+    for (hashval = 0; *s != '\0'; s++)
+        hashval = *s + 31 * hashval;
+
+    return hashval % GRAPH_VERTICES_TABLE_SIZE;
```

Comment:
why do you need it?

Reply:
To get a correct index in vertices_table.
As table size is 20, hashval is divided by 20 and thus vertices_table boundaries are taken into account.
Otherwise hashval may be the number that beyond vertices_table, namely index 19.